### PR TITLE
Add `pref-key` to FML feature variables

### PR DESCRIPTION
**Pull request by jhugman**
Fixes [EXP-2822](https://mozilla-hub.atlassian.net/browse/EXP-2822).

This PR adds to the FML a pref key to feature variables. For example

```yaml
features:
  my-feature:
    variables:
      is-enabled:
        type: Boolean
        default: false
        pref-key: "my-feature.enabled"
```

The `NimbusBuilder` now accepts a `userDefaults` or `sharedPreferences` which stores the prefs.

The generated code for `myFeature.isEnabled` now checks the prefs for a key `my-feature.enabled`.

This allows app code to toggle on or off features, via a pref rather than just experiment.

There are several restrictions for this feature:

- only top level variables in a feature are supported— nesting gets tricky quickly, so we need to see if this is a feature that is actually used.
- only `Boolean`, `Int`, `String`, and `Text` types are supported.

To detect if a user's preference has been set, a new`isModified()` method is provided. This returns `true` if any prefs in the feature have been set.

When `isModified` is `true`, exposure events are suppressed. 

NB. It's not clear (without seeing usage) if this suppressing exposure events is enough before the prefs can be changed by a general population. Alternatives may be to unenroll from any experiments involving the feature. Until this is clear, we advise only exposing such user preferences in a secret settings context.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

**Review by jhugman — COMMENTED** [review]
Self-review

**Review by jeddai — APPROVED** [review]
Overall looks good, I'm loving the refactors and tests. I ran into issues trying to test it in Fenix and didn't want to hold up the review process any longer so I'm approving now. Since it's not a breaking change I'm not particularly concerned about merging it now to get builds going with it and making changes again later if we need to.

File: components/nimbus/ios/Nimbus/FeatureHolder.swift
Comment (jhugman) [reply components/nimbus/ios/Nimbus/FeatureHolder.swift:152]: Adding `public` gives:

:x: error: `public` modified cannot be used in protocols.

File: components/nimbus/ios/Nimbus/FeatureHolder.swift
Comment (jhugman) [inline components/nimbus/ios/Nimbus/FeatureHolder.swift:156]: Adding `public` gives:

⚠️ `public` modifier is redundant for instance method declared in a public extension.

File: components/nimbus/ios/Nimbus/FeatureHolder.swift
Comment (jeddai) [reply components/nimbus/ios/Nimbus/FeatureHolder.swift:152]: Cool I just wasn't sure — iirc we had a few prs in recent history just adding privacy modifiers

File: components/support/nimbus-fml/test/pref_overrides.kts
Comment (jhugman) [reply components/support/nimbus-fml/test/pref_overrides.kts:1]: Please file an issue! I agree, it's getting unwieldy adding these to `workflows.rs` and a `kts` and `swift` file.

**Pull request merged**
